016-database-adapters.md

```diff
@@ -0,0 +1,337 @@
+* GET /v1/database/collections/:id/attributes/:attribute (list attributes)
+* GET /v1/database/collections/:id/attributes/:attribute (get attribute)
+* DELETE /v1/database/collections/:id/attributes/:attribute (delete attribute)
+
+**Collection Indices/Indexes** (no update)
+
+* POST /v1/database/collections/:id/indexes (create index)
+* GET /v1/database/collections/:id/indexes/:index (list indices)
+* GET /v1/database/collections/:id/indexes/:index (get index)
+* DELETE /v1/database/collections/:id/indexes/:index (delete index)
+
+**Collection Documents**
+
+* POST /v1/database/collections/:id/documents (create document)
+* GET /v1/database/collections/:id/documents/:document (list documents)
+* GET /v1/database/collections/:id/documents/:document (get document)
+* PUT /v1/database/collections/:id/documents/:document (update document)
+* DELETE /v1/database/collections/:id/documents/:document (delete document)
+
+> We'll use index and indexes for naming conventions, as its is more consistent than indexs vs indices.
+
+### Database Worker
+
+### Documentation
+
+Below is a list of some of the main topics we should cover as part of the database documnetation. This is in addition to the auto-generated API spec.
+
+* Overview (concept, architecture)
+* Data Types
+* Permissions
+* Operations (queries + examples)
+* Attributes (rules) and indexes
+* Security
+* Benchmarks
+* Performance Tips
+* Known Limitations
+
+### Resources
+
+* SQL Optimization - https://www.eversql.com/
+* Function syntax parsing https://gist.github.com/lamberta/3768814
+
+### Prior art
+
+[prior-art]: #prior-art
+
+<!--
+
+Discuss prior art, both the good and the bad, in relation to this proposal. A
+few examples of what this can include are:
+
+- Does this functionality exist in other software and what experience has their
+  community had?
+- For other teams: What lessons can we learn from what other communities have
+  done here?
+- Papers: Are there any published papers or great posts that discuss this? If
+  you have some relevant papers to refer to, this can serve as a more detailed
+  theoretical background.
+
+This section is intended to encourage you as an author to think about the
+lessons from other software, provide readers of your RFC with a fuller picture.
+If there is no prior art, that is fine - your ideas are interesting to us
+whether they are brand new or if it is an adaptation from other software.
+
+Write your answer below.
+-->
+
+### Unresolved questions
+
+[unresolved-questions]: #unresolved-questions
+
+<!-- What parts of the design do you expect to resolve through the RFC process before this gets merged? -->
+
+<!-- Write your answer below. -->
+
+#### Query Syntax for Filter
+
+Our current query syntax is limted and don't allows us to detect given value types, or to use operators like `OR` and `AND`.
+
+**Current**:
+
+`movie.director.name=Michael Bay`
+
+**New Alternative**:
+
+`movie.director.name.equal('Michael Bay')`
```

Comment:
Agree, this looks good. We could take a look at Mango Queries for more examples:

**.lt()** Match fields “less than” this one.
**.gt()** Match fields “greater than” this one.
**.lte()** Match fields “less than or equal to” this one.
**.gte()** Match fields “greater than or equal to” this one.
**.eq()** Match fields equal to this one.
**.ne()** Match fields not equal to this one.
**.exists()** True if the field should exist, false otherwise.
**.in()** The document field must be set (optional rules).
**.nin()** The document field must not be set (optional rules).
**.or()** Matches if any of the selectors in the given array match. 
**.nor()** Matches if none of the selectors in the given array match.
**.regex()** A regular expression pattern to match against the document field.

Reply:
We could also pass them like this over the REST API, which makes it a bit easier to parse:

`?filter[movie.director.name][equal]=Michael+Bay&filter[limit]=3`

Reply:
For those, like me, who thought Mango was just a misspelled Mongo: 
https://pouchdb.com/guides/mango-queries.html

Reply:
> ?filter[movie.director.name][equal]=Michael+Bay&filter[limit]=3

@TorstenDittmann this does look nice, but will be complex to represent in our SDKs.

Reply:
Will it make sense to allow multiple values like this: `movie.director.name.equal('Michael Bay', 'Steven Spielberg')` instead of having an `OR` operator?

Reply:
> > ?filter[movie.director.name][equal]=Michael+Bay&filter[limit]=3
> 
> @TorstenDittmann this does look nice, but will be complex to represent in our SDKs.

How so?

For example the Web SDK expects an object for the filter with following structure:

```json
{
   "movie.director.name": {
      "equal": [
         "Michael Bay",
         "Steven Spielberg"
      ]
   },
   "movie.year: {
      "equal": 1989
   }
}
```

This may seem very complex for easy queries, but the filter is in a readable format for computers and the IDE.

This allows easy changes to the filter object with code and autocomplete in the IDE.

Reply:
If we allow devs to pass strings like `movie.director.name.equal('Michael Bay')` - there will be no autocomplete since we cannot offer typings for this.

Reply:
Example can be found [here](https://www.typescriptlang.org/play?#code/KYOwrgtgBA8gDsATgQwC4HtEGcoG8BQURUAogI5jIA2UAvFAOTAXUMA0hxAcuquZTXoMQvfqw7EoAcUTA0SOowDms+YnaciAGWBYsCoVV371HAL758qAJ4JYCFBmwAVW8EUFJAbQAKUAJYg9khomFgAugD8AFxQAEbo6EbIIAA+WKiIgUrpmdle4angEHFIRZCliAX4FlZuUABi-lSoSFiudvSexF5wiOgONrEZWSBK4bHwIU7tbjWWAPQLUFqJANZQABZI7pYAxuggGVAAZs2t2LFNLW0d7l2aUABEEOgAbv7AAHQAJv6yeycXxAyAgwCesW6kiIT2YAghUC8j2hxCeAFl-HtNshgDQAELIaxPCQo6FPADKrTeoCg5LgnyolSUT2RxHCjzMJJhrw+32sckQCKh0JUcgusQATAAGCUARg5XOePM+X1Q-lQRiFrKISygzkQ1igyDAGAOEDgRlaWx2HJqQA)

Reply:
It would be very hard to describe this type of un-structured data as part of the REST and GraphQL APIs

Reply:
This is actually doable and also already done by other tools. For example Directus does a similiar approach:
https://docs.directus.io/reference/api/query/#filter

But I found a blocker for my approach, OpenAPI/Swagger only allows objects in query params without nesting:
https://github.com/OAI/OpenAPI-Specification/issues/1706#issuecomment-458949282